### PR TITLE
Fix NullPointerException in canAccessParticipation

### PR DESCRIPTION
**Pull request by MaisiKoleni**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemis-staging.ase.in.tum.de and https://artemistest2.ase.in.tum.de/ with ~50 exercise details pages.
- [x] Server: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
Fix occasional `NullPointerException` when checking participation access with `canAccessParticipation`. This relates to Sentry issue 
`ARTEMIS-3YH`, see https://sentry.ase.in.tum.de/share/issue/b0e500ac54f74df9b402bf9887acdbb1/.

### Description
Rewrite logic for fetching the exercise in case it is missing, we directly fetch the exercise now from the DB. An added benefit is, the code is simpler and more readable.

### Steps for Testing
1. Activate the developer tools in your browser and look into the network tab
2. Open programming exercise details of various exercises and after each check the network tab. (This does not happen all the time, I would recommend trying exercises with different: age, course, exam, ...)
3. There should be no server error related to the latest results or submissions.
4. It might be a good idea to check sentry as well that the linked issue did not occur during testing in the test environment.

### Review Progress

- Code Review
  - [x] Review 1
  - [x] Review 2
- Manual Tests
  - [x] Test 1
  - [x] Test 2

### Test Coverage

* ProgrammingExerciseParticipationService.java: 95%
  image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java [hunk 1]
```diff
@@ -172,9 +171,14 @@
...
-
-    ProgrammingExercise findOneBySolutionParticipationId(Long solutionParticipationId);
+    @Query("""
+            SELECT pe FROM ProgrammingExercise pe
+            LEFT JOIN pe.studentParticipations pep
+            WHERE pep.id = :#{#participationId}
+                OR pe.templateParticipation.id = :#{#participationId}
+                OR pe.solutionParticipation.id = :#{#participationId}
+            """)
+    Optional<ProgrammingExercise> findByParticipationId(Long participationId);
```
Comment (MerlinMehmed) [inline src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:181]: The guideline for SQL query params (https://docs.artemis.ase.in.tum.de/dev/guidelines/server/#proper-annotation-of-sql-query-parameters) suggest adding @ Param("participationId") before the parameter name

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java [hunk 2]
```diff
@@ -453,23 +457,13 @@ SELECT COUNT (DISTINCT p) FROM ProgrammingExerciseStudentParticipation p
...
-     * @param participation The template participation
-     * @return The ProgrammingExercise where the given Participation is the template Participation
-     */
-    default ProgrammingExercise getExercise(TemplateProgrammingExerciseParticipation participation) {
-        return findOneByTemplateParticipationId(participation.getId());
-    }
-
-    /**
-     * Find the ProgrammingExercise where the given Participation is the solution Participation
+     * Find the ProgrammingExercise where the given Participation is either a student, template or solution Participation
```
Comment (rs-maker) [inline src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:460]: Please change this comment to be more clear.

**Review by rs-maker — APPROVED** [review]
Looks good left a small comment.

**Review by rs-maker — APPROVED** [review]
Tried loading 10 different programming exercises and didnt get the error so i think its safe to assume that this at least helped.

So both code and testing reapproval.

**Review by Malyuk-A — DISMISSED** [review]
Tested during Testing Session on TS1. Works as expected.

**Review by MerlinMehmed — APPROVED** [review]
Tested in a testing session with many exercises and the error that was previously occurring didn't occur. 
Code changes also look good.

**Pull request merged**